### DISCUSSION TITLE
[MagickWand]: Access to pixels from external library (wxWidgets).

### DISCUSSION THREAD

**Discussion by DebugBSD** [ImageMagick/ImageMagick#2967] [Development] [answered]
Hi, good morning.

This is my first post so I'm sorry if this is not the correct place to ask this question. Point me the correct place if it's the case please.

My working on a kind of World Editor for my 2D game engine and I'm using wxWidgets as GUI. The thing is, I'm using MagickWand C API to open all images, process them (Crop, fade, transform, ...) and everything works fine. The problem comes when using huge textures (4096x4096 pixels). I'm using a lot of textures as layers and when I try to extract subregions of that textures or even when I try to open that textures, the program uses a lot of Memory and takes a lot of time so I wonder if I'm doing correctly.

This is the fragment of code used to read the pixels from the image:

void weImage::readPixelsFromImage()
{
    PixelWand *color;

    MagickBooleanType status;

    wxSize size = getSize();

    m_pPixels = (uint8_t*)malloc((size_t)size.x * (size_t)size.y * (size_t)3);
    m_pAlphaPixels = (uint8_t*)malloc((size_t)size.x * (size_t)size.y);

    char* pixels = (char *)malloc((size_t)size.x * (size_t)size.y * (size_t)4);

    MagickBooleanType isOK = MagickExportImagePixels(m_pMagickWand, 0,0, size.x, size.y, "RGBA", CharPixel, pixels);
    if (isOK == MagickTrue)
    {
        if (pixels != NULL)
        {
            int pixelsIndex = 0;
            int alphaIndex = 0;
            for (int i = 0; i < size.x * size.y * 4; i += 4)
            {
                m_pPixels[pixelsIndex++] = (uint8_t)pixels[i];
                m_pPixels[pixelsIndex++] = (uint8_t)pixels[i + 1];
                m_pPixels[pixelsIndex++] = (uint8_t)pixels[i + 2];
                m_pAlphaPixels[alphaIndex++] = (uint8_t)pixels[i + 3];
            }
            std::free(pixels);
        }
    }
    else
    {
        // Log error...

        // Clean resources
        if (m_pPixels != NULL)
        {
            std::free(m_pPixels);
            m_pPixels = NULL;
        }

        if (m_pAlphaPixels != NULL)
        {
            std::free(m_pAlphaPixels);
            m_pAlphaPixels = NULL;
        }

        if (pixels != NULL)
        {
            std::free(pixels);
        }
    }
}

The thing is, the program works perfectly but the problem is that fragmen which consumes a lot of memory and time. So my question is: Is that fragment a correct one? Or maybe there is a better option to read such pixels from the image. 

I tried to search on the internet about more documentation about the best way to access the pixels but, the only thing I've found is something related to Pixel Iterator (https://imagemagick.org/api/pixel-iterator.php) which I don't understand and I couldn't find a good example about using it.

Thanks for everything!
Guille

**Comment by snibgo** [comment]
IM source code includes example *.c programs that use PixelIterator.

**Reply by DebugBSD** [reply]
Thanks for the answer.

I'm sorry if I haven't explained correctly my problem but, I wanted to say if the way I've used the API to extract the pixels from the image is the correct or not. I wasn't asking anything about Pixel Iterator. I just said that Pixel Iterator seems to be related to my problem related to extract pixels from the image but I wasn't sure about that. 

Anyway, I'll look at those examples to see if there is a better way to extract the pixels. 
Thanks!
Guille

**Comment by snibgo** [comment]
There are alternatives:

1. You could directly access pixel values in m_pMagickWand. This would save memory and time.

2. After calling MagickExportImagePixels(), you could access pixel values in pixels[] array. This would save time.

3. If you really want m_pPixels[] and m_pAlphaPixels, you could parallelize the copying. This would save time.

EDIT: You hard-code a limitation of 8 bits per (integer) pixel value. This may suit your current purposes, of course, but you might consider making your code more adaptable should you wish to use 16 bits or floating point in the future.

**Reply by DebugBSD** [reply]
Thank you so much! Thats was really helpful.

Now, if I choose the option 1 (access to pixels through m_pMagickWand), whats is the way to procced? Should I use Pixel Iterator in this case or is there a better option/function? Please, can you point me to the correct direction (maybe a web page, or an example or something like that) ?

NOTE: If I'm using 8 bits per pixel it's because I didn't know how to use 16bits. I'm quite beginner in the ImageMagick framework as you could see. I'll take note about using 16bits when I'm pretty sure about what I'm doing.

Thank you so much!
Guille

**Comment by snibgo** [comment, marked as answer]
Most of my work uses the MagickCore interface, not MagickWand. But the MagickWand pixel iterators are fairly simple, and an efficient way to access pixels for serial use (eg walking through all the pixels).

For example, call NewPixelIterator(magick_wand). Then call PixelSetIteratorRow() for the row number you want, then PixelGetCurrentIteratorRow() to get an array of pixels for that row, then PixelGetMagickColor() to access an individual pixel. If you need to loop through all the rows and have OpenMP, you can do this in parallel.

However, if you want random access to the pixels, MagickGetImagePixelColor() and MagickSetImagePixelColor() are useful. These are simpler, but less efficient when used for serial access as they create and destroy a cache view for every pixel.

For any processing you need, you may consider whether IM contains functions that do the task so you don't need to write it yourself. For example, if you need to multiply pixel values by a factor, using MagickEvaluateImage() will be easier than writing your own code, and will execute faster.


MagickWand functions are documented at https://imagemagick.org/api/magick-image.php . However, that web page is created from comments in the code, which may not be accurate. I prefer to read the code itself, and take the comments with a pinch of salt (ie assume they may not be correct).

**Reply by DebugBSD** [reply]
Thank you so much! I think this solves my problem. 

I'll do the way you told me (Using NewPixelIterator(), PixelSetIteratorRow(), ...) 

Regarding to the documentation I understant that problem. It's poorly documented. The thing is, I don't want to get into ImageMagick source code if I don't need to but, I have to say that you are right! I think that finally I'll start reading the source code instead of the documentation. It's a shame because I love this library. It's amazing how many things you can do with this library. I think that a good api , good documentation and good examples about how to use it, could let more programmers to use the library, Even programmers (like me) who doesn't have any previous experience with image processing.

Again, thank you so much.
Guille
